internal/catalog/internal/controllers/failover/controller.go

```diff
@@ -0,0 +1,301 @@
+
+	_, err = rt.Client.WriteStatus(ctx, &pbresource.WriteStatusRequest{
+		Id:     failoverPolicy.Resource.Id,
+		Key:    StatusKey,
+		Status: newStatus,
+	})
+
+	if err != nil {
+		rt.Logger.Error("error encountered when attempting to update the resource's failover policy status", "error", err)
+		return err
+	}
+
+	rt.Logger.Trace("resource's failover policy status was updated",
+		"conditions", newStatus.Conditions)
+	return nil
+}
+
+func getFailoverPolicy(ctx context.Context, rt controller.Runtime, id *pbresource.ID) (*resource.DecodedResource[pbcatalog.FailoverPolicy, *pbcatalog.FailoverPolicy], error) {
+	res, err := getResource(ctx, rt, id)
+	if err != nil {
+		return nil, err
+	}
+	if res == nil {
+		return nil, nil
+	}
+
+	return resource.Decode[pbcatalog.FailoverPolicy, *pbcatalog.FailoverPolicy](res)
+}
+
+func getService(ctx context.Context, rt controller.Runtime, id *pbresource.ID) (*resource.DecodedResource[pbcatalog.Service, *pbcatalog.Service], error) {
+	res, err := getResource(ctx, rt, id)
+	if err != nil {
+		return nil, err
+	}
+	if res == nil {
+		return nil, nil
+	}
+
+	return resource.Decode[pbcatalog.Service, *pbcatalog.Service](res)
+}
+
+func getResource(ctx context.Context, rt controller.Runtime, id *pbresource.ID) (*pbresource.Resource, error) {
+	rsp, err := rt.Client.Read(ctx, &pbresource.ReadRequest{Id: id})
+	switch {
+	case status.Code(err) == codes.NotFound:
+		return nil, nil
+	case err != nil:
+		return nil, err
+	}
+	return rsp.Resource, nil
+}
+
+func computeNewStatus(
+	failoverPolicy *resource.DecodedResource[pbcatalog.FailoverPolicy, *pbcatalog.FailoverPolicy],
+	service *resource.DecodedResource[pbcatalog.Service, *pbcatalog.Service],
+	otherServices map[resource.ReferenceKey]*resource.DecodedResource[pbcatalog.Service, *pbcatalog.Service],
+) *pbresource.Status {
+	if service == nil {
+		return &pbresource.Status{
+			ObservedGeneration: failoverPolicy.Resource.Generation,
+			Conditions: []*pbresource.Condition{
+				ConditionMissingService,
+			},
+		}
+	}
+
+	allowedPortProtocols := make(map[string]pbcatalog.Protocol)
+	for _, port := range service.Data.Ports {
+		if port.Protocol == pbcatalog.Protocol_PROTOCOL_MESH {
+			continue // skip
+		}
+		allowedPortProtocols[port.TargetPort] = port.Protocol
+	}
+
+	serviceHasPort := func(dest *pbcatalog.FailoverDestination) *pbresource.Condition {
+		key := resource.NewReferenceKey(dest.Ref)
+		if destService, ok := otherServices[key]; ok {
+			found := false
+			mesh := false
+			for _, port := range destService.Data.Ports {
+				if port.TargetPort == dest.Port {
+					found = true
+					if port.Protocol == pbcatalog.Protocol_PROTOCOL_MESH {
+						mesh = true
+					}
+					break
+				}
+			}
+
+			if !found {
+				return ConditionUnknownDestinationPort(dest.Ref, dest.Port)
+			} else if mesh {
+				return ConditionUsingMeshDestinationPort(dest.Ref, dest.Port)
+			}
+		} else {
+			return ConditionMissingDestinationService(dest.Ref)
+		}
+		return nil
+	}
```

Comment:
Would it be difficult to extract this out into its own func instead of inlined and relying on closures? If its not I think it would result in slightly easier to grok method here and allow testing of this in isolation if we wanted.

Reply:
done